AzureStorageCleanupThirdPartyTests.testCleanup failures on CI

**Build scan**: https://gradle-enterprise.elastic.co/s/ftwfuwjjk5nv4

**Repro line**: `./gradlew ':plugins:repository-azure:azureThirdPartyTest' --tests "org.elasticsearch.repositories.azure.AzureStorageCleanupThirdPartyTests.testCleanup" -Dtests.seed=60997A36A0D370AE -Dtests.security.manager=true -Dtests.locale=hi -Dtests.timezone=America/Lima -Druntime.java=8`

**Reproduces locally?**: No

**Applicable branches**: `master`, `7.x`, `7.11`

**Failure history**: [build stats](https://build-stats.elastic.co/app/kibana#/discover?_g=(refreshInterval:(pause:!t,value:0),time:(from:now-30d,mode:quick,to:now))&_a=(columns:!(branch,message),index:e58bf320-7efd-11e8-bf69-63c8ef516157,interval:auto,query:(language:lucene,query:'test:testCleanup'),sort:!(time,desc)))
First failure occurred `2020-12-17T00:03:45.235Z` on `7.11`.
**Failure excerpt**:
```
java.lang.AssertionError: java.lang.AssertionError: 	
Expected: iterable with items [] in any order	
     but: no match for: "foo"	
	at __randomizedtesting.SeedInfo.seed([60997A36A0D370AE:30B32998D97EE01B]:0)	
	at org.elasticsearch.repositories.blobstore.BlobStoreTestUtil.assertConsistency(BlobStoreTestUtil.java:136)	
	at org.elasticsearch.repositories.AbstractThirdPartyRepositoryTestCase.assertConsistentRepository(AbstractThirdPartyRepositoryTestCase.java:260)	
	at org.elasticsearch.repositories.AbstractThirdPartyRepositoryTestCase.testCleanup(AbstractThirdPartyRepositoryTestCase.java:218)	
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)	
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)	
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)	
	at java.lang.reflect.Method.invoke(Method.java:498)	
	at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1750)	
	at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:938)	
	at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:974)	
	at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:988)	
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)	
	at org.apache.lucene.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:49)	
	at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)	
	at org.apache.lucene.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:48)	
	at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)	
	at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)	
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)	
	at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368)	
	at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:817)	
	at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:468)	
	at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:947)	
	at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:832)	
	at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:883)	
	at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:894)	
	at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)	
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)	
	at org.apache.lucene.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:41)	
	at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)	
	at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)	
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)	
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)	
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)	
	at org.apache.lucene.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)	
	at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)	
	at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)	
	at org.apache.lucene.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:54)	
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)	
	at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368)	
	at java.lang.Thread.run(Thread.java:748)	
Caused by: java.lang.AssertionError: 	
Expected: iterable with items [] in any order	
     but: no match for: "foo"	
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)	
	at org.junit.Assert.assertThat(Assert.java:956)	
	at org.junit.Assert.assertThat(Assert.java:923)	
	at org.elasticsearch.repositories.blobstore.BlobStoreTestUtil.assertSnapshotUUIDs(BlobStoreTestUtil.java:207)	
	at org.elasticsearch.repositories.blobstore.BlobStoreTestUtil.lambda$assertConsistency$0(BlobStoreTestUtil.java:127)	
	at org.elasticsearch.action.ActionRunnable.lambda$supply$0(ActionRunnable.java:58)	
	at org.elasticsearch.action.ActionRunnable$2.doRun(ActionRunnable.java:73)	
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:743)	
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)	
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)	
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)	
	at java.lang.Thread.run(Thread.java:748)
```

Mute incoming.

--- a/plugins/repository-azure/src/internalClusterTest/java/org/elasticsearch/repositories/azure/AzureBlobStoreRepositoryTests.java
+++ b/plugins/repository-azure/src/internalClusterTest/java/org/elasticsearch/repositories/azure/AzureBlobStoreRepositoryTests.java
@@ -49,7 +49,9 @@ import java.util.Map;
 import java.util.function.Predicate;
 import java.util.regex.Pattern;
 
+import static org.hamcrest.Matchers.anEmptyMap;
 import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.is;
 
 @SuppressForbidden(reason = "this test uses a HttpServer to emulate an Azure endpoint")
 public class AzureBlobStoreRepositoryTests extends ESMockAPIBasedRepositoryIntegTestCase {
@@ -253,7 +255,7 @@ public class AzureBlobStoreRepositoryTests extends ESMockAPIBasedRepositoryInteg
 
             Randomness.shuffle(blobsToDelete);
             container.deleteBlobsIgnoringIfNotExists(blobsToDelete);
-            assertThat(container.listBlobs().size(), equalTo(0));
+            assertThat(container.listBlobs(), is(anEmptyMap()));
         }
     }
 }

--- a/plugins/repository-azure/src/internalClusterTest/java/org/elasticsearch/repositories/azure/AzureBlobStoreRepositoryTests.java
+++ b/plugins/repository-azure/src/internalClusterTest/java/org/elasticsearch/repositories/azure/AzureBlobStoreRepositoryTests.java
@@ -24,6 +24,7 @@ import com.sun.net.httpserver.Headers;
 import com.sun.net.httpserver.HttpExchange;
 import com.sun.net.httpserver.HttpHandler;
 import fixture.azure.AzureHttpHandler;
+import org.elasticsearch.common.Randomness;
 import org.elasticsearch.common.SuppressForbidden;
 import org.elasticsearch.common.blobstore.BlobContainer;
 import org.elasticsearch.common.blobstore.BlobPath;
@@ -39,9 +40,11 @@ import org.elasticsearch.rest.RestStatus;
 
 import java.io.IOException;
 import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
 import java.util.Base64;
 import java.util.Collection;
 import java.util.Collections;
+import java.util.List;
 import java.util.Map;
 import java.util.function.Predicate;
 import java.util.regex.Pattern;
@@ -228,6 +231,28 @@ public class AzureBlobStoreRepositoryTests extends ESMockAPIBasedRepositoryInteg
             }
 
             container.delete();
+            assertThat(container.listBlobs().size(), equalTo(0));
+        }
+    }
+
+    public void testDeleteBlobsIgnoringIfNotExists() throws Exception {
+        try (BlobStore store = newBlobStore()) {
+            final BlobContainer container = store.blobContainer(new BlobPath());
+            List<String> blobsToDelete = new ArrayList<>();
+            for (int i = 0; i < 10; i++) {
+                byte[] bytes = randomBytes(randomInt(100));
+                String blobName = randomAlphaOfLength(10);
+                container.writeBlob(blobName, new BytesArray(bytes), false);
+                blobsToDelete.add(blobName);
+            }
+
+            // Try to delete non existent blobs
+            for (int i = 0; i < 10; i++) {
+                blobsToDelete.add(randomName());
+            }
+
+            Randomness.shuffle(blobsToDelete);
+            container.deleteBlobsIgnoringIfNotExists(blobsToDelete);
             assertThat(container.listBlobs().size(), equalTo(0));
         }
     }

--- a/plugins/repository-azure/src/internalClusterTest/java/org/elasticsearch/repositories/azure/AzureStorageCleanupThirdPartyTests.java
+++ b/plugins/repository-azure/src/internalClusterTest/java/org/elasticsearch/repositories/azure/AzureStorageCleanupThirdPartyTests.java
@@ -122,11 +122,4 @@ public class AzureStorageCleanupThirdPartyTests extends AbstractThirdPartyReposi
         }));
         future.actionGet();
     }
-
-    @AwaitsFix(bugUrl = "https://github.com/elastic/elasticsearch/issues/66633")
-    @Override
-    // This override is only here so we can mute the test without muting the whole suite, remove it when the test is fixed
-    public void testCleanup() throws Exception {
-        super.testCleanup();
-    }
 }

--- a/plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java
+++ b/plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java
@@ -275,24 +275,25 @@ public class AzureBlobStore implements BlobStore {
                 List<Mono<Void>> deleteTasks = new ArrayList<>(blobs.size());
                 final BlobContainerAsyncClient blobContainerClient = asyncClient.getBlobContainerAsyncClient(container);
                 for (String blob : blobs) {
-                    deleteTasks.add(blobContainerClient.getBlobAsyncClient(blob).delete());
+                    final Mono<Void> deleteTask = blobContainerClient.getBlobAsyncClient(blob)
+                        .delete()
+                        // Ignore not found blobs
+                        .onErrorResume(e -> (e instanceof BlobStorageException) && ((BlobStorageException) e).getStatusCode() == 404,
+                            throwable -> Mono.empty());
+                    deleteTasks.add(deleteTask);
                 }
 
                 executeDeleteTasks(deleteTasks);
             });
-        } catch (BlobStorageException e) {
-            if (e.getStatusCode() != 404) {
-                throw new IOException("Unable to delete blobs " + blobs, e);
-            }
         } catch (Exception e) {
             throw new IOException("Unable to delete blobs " + blobs, e);
         }
     }
 
     private void executeDeleteTasks(List<Mono<Void>> deleteTasks) {
-        Flux.merge(deleteTasks)
-            .collectList()
-            .block();
+        // zipDelayError executes all tasks in parallel and delays
+        // error propagation until all tasks have finished.
+        Mono.zipDelayError(deleteTasks, results -> null).block();
     }
 
     public InputStream getInputStream(String blob, long position, final @Nullable Long length) throws IOException {

--- a/test/fixtures/azure-fixture/src/main/java/fixture/azure/AzureHttpHandler.java
+++ b/test/fixtures/azure-fixture/src/main/java/fixture/azure/AzureHttpHandler.java
@@ -152,8 +152,12 @@ public class AzureHttpHandler implements HttpHandler {
 
             } else if (Regex.simpleMatch("DELETE /" + account + "/" + container + "/*", request)) {
                 // Delete Blob (https://docs.microsoft.com/en-us/rest/api/storageservices/delete-blob)
-                blobs.entrySet().removeIf(blob -> blob.getKey().startsWith(exchange.getRequestURI().getPath()));
-                exchange.sendResponseHeaders(RestStatus.ACCEPTED.getStatus(), -1);
+                final boolean deleted = blobs.entrySet().removeIf(blob -> blob.getKey().startsWith(exchange.getRequestURI().getPath()));
+                if (deleted) {
+                    exchange.sendResponseHeaders(RestStatus.ACCEPTED.getStatus(), -1);
+                } else {
+                    exchange.sendResponseHeaders(RestStatus.NOT_FOUND.getStatus(), -1);
+                }
 
             } else if (Regex.simpleMatch("GET /" + account + "/" + container + "?*restype=container*comp=list*", request)) {
                 // List Blobs (https://docs.microsoft.com/en-us/rest/api/storageservices/list-blobs)

